### PR TITLE
Implement concurrency changing

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Part of https://github.com/Particular/DeveloperExperience/issues/1516

### CODE REVIEW

File: src/NServiceBus.Transport.Msmq/MessagePump.cs [hunk 1]
```diff
@@ -85,6 +85,28 @@ public Task StartReceive(CancellationToken cancellationToken = default)
...
+                //Drain and dispose of the old semaphore
+                while (oldLimiter.CurrentCount != oldMaxConcurrency)
+                {
+                    await Task.Delay(50, cancellationToken).ConfigureAwait(false);
+                }
+                oldLimiter.Dispose();
+            }
+            catch (Exception ex) when (ex.IsCausedBy(cancellationToken))
+            {
+                //Ignore, we are stopping anyway
```
Comment (danielmarbach) [inline src/NServiceBus.Transport.Msmq/MessagePump.cs:106]: This comment is not entirely correct in this context

File: src/NServiceBus.Transport.Msmq/MessagePump.cs [hunk 2]
```diff
@@ -85,6 +85,28 @@ public Task StartReceive(CancellationToken cancellationToken = default)
             return Task.CompletedTask;
         }
 
+        public async Task ChangeConcurrency(PushRuntimeSettings newLimitations, CancellationToken cancellationToken = new CancellationToken())
```
Comment (danielmarbach) [inline src/NServiceBus.Transport.Msmq/MessagePump.cs:88]: Why ` = new CancellationToken()`? Shouldn't this be `default`?

File: src/NServiceBus.Transport.Msmq/MessagePump.cs [hunk 3]
```diff
@@ -175,12 +197,14 @@ async Task PumpMessages(CancellationToken messagePumpCancellationToken)
...
-                    await concurrencyLimiter.WaitAsync(messagePumpCancellationToken).ConfigureAwait(false);
+                    var localLimiter = concurrencyLimiter;
+
+                    await localLimiter.WaitAsync(messagePumpCancellationToken).ConfigureAwait(false);
 
                     _ = Task.Factory.StartNew(state =>
                         {
                             var (messagePump, cancellationToken) = ((MessagePump, CancellationToken))state;
-                            return ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(messagePump, cancellationToken);
+                            return ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(messagePump, localLimiter, cancellationToken);
```
Comment (danielmarbach) [inline src/NServiceBus.Transport.Msmq/MessagePump.cs:207]: This is now capturing a closure which was previously avoided. To avoid the closure, you need to pass in the limiter via the value tuple and extract it inside the `StartNew`

But see below comment. Maybe this is not needed

File: src/NServiceBus.Transport.Msmq/MessagePump.cs [hunk 4]
```diff
@@ -192,7 +216,7 @@ async Task PumpMessages(CancellationToken messagePumpCancellationToken)
         }
 
         // This is static to prevent the method from accessing fields in the pump since that causes variable capturing and cause extra allocations
-        static async Task ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(MessagePump messagePump, CancellationToken messagePumpCancellationToken)
+        static async Task ReceiveMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(MessagePump messagePump, SemaphoreSlim localConcurrencyLimiter, CancellationToken messagePumpCancellationToken)
```
Comment (danielmarbach) [inline src/NServiceBus.Transport.Msmq/MessagePump.cs:219]: `localConcurrencyLimiter` maybe is irrelevant. How about `concurrencyLimiter`. Also why is this change needed? Following the logic you could also assign extract it from the pump at the beginning, assign it to a local and release that one at the end couldn't you?

Comment (SzymonPobiega) [reply src/NServiceBus.Transport.Msmq/MessagePump.cs:219]: I think it is needed. We need to release the same semaphore as we acquired in the previous method to make sure we drain all semaphores correctly. So I'll fix the passing of the semaphore to use the tuple.

File: src/NServiceBus.Transport.Msmq/MessagePump.cs (hunk 3 above)
Comment (SzymonPobiega) [reply src/NServiceBus.Transport.Msmq/MessagePump.cs:207]: Thanks, I'll fix it.

File: src/NServiceBus.Transport.Msmq/MessagePump.cs (hunk 2 above)
Comment (SzymonPobiega) [reply src/NServiceBus.Transport.Msmq/MessagePump.cs]: 👍

File: src/NServiceBus.Transport.Msmq/MessagePump.cs (hunk 3 above)
Comment (SzymonPobiega) [reply src/NServiceBus.Transport.Msmq/MessagePump.cs]: 👍

**Review by WojcikMike — APPROVED** [review]
